File: src/facade/examples/imagesearch/ImageSearchFacade.js
'use strict';

const MetadataAdapter = require('../../../adapter/examples/metadata/MetadataAdapter');
const InMemoryCatalogFactory = require('../../../abstractfactory/examples/searchcatalog/InMemoryCatalogFactory');
const SearchIntent = require('../../../strategy/examples/searchintent/SearchIntent');
const AdoptionSearchStrategy = require('../../../strategy/examples/searchintent/AdoptionSearchStrategy');
const SaleSearchStrategy = require('../../../strategy/examples/searchintent/SaleSearchStrategy');
const VerifyLossSearchStrategy = require('../../../strategy/examples/searchintent/VerifyLossSearchStrategy');

/**
 * RF2.1: fachada unica para subir imagen + intent.
 * RNF2.2: mide latencia extremo a extremo.
 */
class ImageSearchFacade {
  constructor(catalogFactory) {
    this.adapter = new MetadataAdapter();
    this.catalogFactory = catalogFactory || new InMemoryCatalogFactory();
  }

  uploadAndSearch(rawImage, intent) {
    if (!intent) throw new Error('Intent requerido (RF2.2)');
    const start = Date.now();
    const meta = this.adapter.adapt(rawImage);
    const strategy = this._pickStrategy(intent);
    const result = strategy.search(meta);
    const elapsed = Date.now() - start;
    console.log(`[ImageSearchFacade] intent=${intent} meta=${meta.toJson()} elapsedMs=${elapsed}`);
    return result;
  }

  _pickStrategy(intent) {
    switch (intent) {
      case SearchIntent.ADOPTION:    return new AdoptionSearchStrategy(this.catalogFactory);
      case SearchIntent.SALE:        return new SaleSearchStrategy(this.catalogFactory);
      case SearchIntent.VERIFY_LOSS: return new VerifyLossSearchStrategy(this.catalogFactory);
      default: throw new Error(`Intent desconocido: ${intent}`);
    }
  }
}

module.exports = ImageSearchFacade;

File: src/facade/examples/imagesearch/ImageSearchFacade.java
package facade.examples.imagesearch;

import abstractfactory.examples.searchcatalog.CatalogFactory;
import abstractfactory.examples.searchcatalog.InMemoryCatalogFactory;
import adapter.examples.metadata.ImageMetadata;
import adapter.examples.metadata.MetadataAdapter;
import adapter.examples.metadata.RawImage;
import strategy.examples.searchintent.AdoptionSearchStrategy;
import strategy.examples.searchintent.SaleSearchStrategy;
import strategy.examples.searchintent.SearchIntent;
import strategy.examples.searchintent.SearchResult;
import strategy.examples.searchintent.SearchStrategy;
import strategy.examples.searchintent.VerifyLossSearchStrategy;

/**
 * RF2.1: single upload interface. Hides adapter + strategy + factory wiring.
 * RNF2.2: response measured; must stay under target window.
 */
public class ImageSearchFacade {

    private final MetadataAdapter adapter = new MetadataAdapter();
    private final CatalogFactory catalogFactory;

    public ImageSearchFacade() { this(new InMemoryCatalogFactory()); }

    public ImageSearchFacade(CatalogFactory catalogFactory) {
        this.catalogFactory = catalogFactory;
    }

    public SearchResult uploadAndSearch(RawImage rawImage, SearchIntent intent) {
        if (intent == null) throw new IllegalArgumentException("Intent required (RF2.2)");
        long start = System.currentTimeMillis();

        ImageMetadata meta = adapter.adapt(rawImage);
        SearchStrategy strategy = pickStrategy(intent);
        SearchResult result = strategy.search(meta);

        long elapsed = System.currentTimeMillis() - start;
        System.out.println("[ImageSearchFacade] intent=" + intent
                + " meta=" + meta.toJson() + " elapsedMs=" + elapsed);
        return result;
    }

    private SearchStrategy pickStrategy(SearchIntent intent) {
        switch (intent) {
            case ADOPTION:    return new AdoptionSearchStrategy(catalogFactory);
            case SALE:        return new SaleSearchStrategy(catalogFactory);
            case VERIFY_LOSS: return new VerifyLossSearchStrategy(catalogFactory);
            default: throw new IllegalStateException("Unknown intent: " + intent);
        }
    }
}
